### PR TITLE
[INV-37] Add login page & logic

### PR DESCRIPTION
**Pull request by ShahobiddinIsmoilov**
https://github.com/user-attachments/assets/90b6b0b7-8738-4dd4-a0c1-78f1ae770677

### CODE REVIEW

File: src/hooks/auth.hook.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import { type LoginResponse } from '../modules/auth/login/types/login-response.type';
+
+export function useAuth() {
+  const dispatch = useDispatch();
+  const currentUser = useSelector((state: StoreState) => state.auth.currentUser);
+  const status = useSelector((state: StoreState) => state.auth.status);
+  const error = useSelector((state: StoreState) => state.auth.error);
+
+  async function login(credentials: { email: string; password: string }) {
+    const baseURL = import.meta.env.VITE_API_BASE_URL;
```
Comment (NegatiV-64) [inline src/hooks/auth.hook.tsx]: I'd suggest to create an api client using axios instance

File: src/hooks/auth.hook.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    dispatch(setStatus('pending'));
+
+    try {
+      const response = await fetch(`${baseURL}/auth/login`, {
```
Comment (NegatiV-64) [inline src/hooks/auth.hook.tsx]: Please use Tanstack Query

File: src/hooks/auth.hook.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      const responseJson: LoginResponse = await response.json();
+      dispatch(
+        setUser({
+          accessToken: responseJson.accessToken,
+          refreshToken: responseJson.refreshToken,
+        })
+      );
+    } catch (err) {
+      const errorMessage =
+        err instanceof Error ? err.message : 'An unexpected error occurred while trying to log in. Please try again';
```
Comment (NegatiV-64) [inline src/hooks/auth.hook.tsx]: Please write an util function to parse the errors

File: src/hooks/auth.hook.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import { useSelector, useDispatch } from 'react-redux';
+import { setUser, clearUser, setError, setStatus } from '#/store/slices/auth-slice';
+import { type StoreState } from '#/store/store';
+import { type LoginResponse } from '../modules/auth/login/types/login-response.type';
+
+export function useAuth() {
+  const dispatch = useDispatch();
+  const currentUser = useSelector((state: StoreState) => state.auth.currentUser);
+  const status = useSelector((state: StoreState) => state.auth.status);
+  const error = useSelector((state: StoreState) => state.auth.error);
```
Comment (NegatiV-64) [inline src/hooks/auth.hook.tsx]: Add logic for page refreshes

File: src/modules/auth/login/components/register-link.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import Box from '@mui/material/Box';
+import Link from '@mui/material/Link';
+import Typography from '@mui/material/Typography';
+import { styles } from '../styles/styles';
```
Comment (NegatiV-64) [inline src/modules/auth/login/components/register-link.tsx]: Please move the styles related to this component, to this component

File: src/modules/auth/login/forms/login-form.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import TextField from '@mui/material/TextField';
+import Stack from '@mui/material/Stack';
+import Typography from '@mui/material/Typography';
+import Button from '@mui/material/Button';
+import Box from '@mui/material/Box';
+import Link from '@mui/material/Link';
+import { styles } from '../styles/styles';
```
Comment (NegatiV-64) [inline src/modules/auth/login/forms/login-form.tsx]: Please move the styles related to this component, to this component

File: src/modules/auth/login/forms/login-form.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  } = useForm<LoginFormData>({
+    resolver: zodResolver(loginFormDataSchema),
+  });
+
+  async function onSubmit(loginFormData: LoginFormData) {
+    await login({ email: loginFormData.email, password: loginFormData.password });
+  }
+
+  useEffect(() => {
+    if (status === 'success') {
```
Comment (NegatiV-64) [inline src/modules/auth/login/forms/login-form.tsx]: Remove the useEffect

File: src/modules/auth/login/forms/login-form.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+  } = useForm<LoginFormData>({
+    resolver: zodResolver(loginFormDataSchema),
+  });
+
+  async function onSubmit(loginFormData: LoginFormData) {
+    await login({ email: loginFormData.email, password: loginFormData.password });
```
Comment (NegatiV-64) [inline src/modules/auth/login/forms/login-form.tsx]: Handle the redirect logic in this place

File: src/modules/auth/login/forms/login-form.tsx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      </Typography>
+
+      {status === 'failed' && (
+        <Typography color="error" align="center" mt={1}>
+          {error || 'Login failed. Please try again.'}
+        </Typography>
+      )}
+
+      <form
+        // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (NegatiV-64) [inline src/modules/auth/login/forms/login-form.tsx]: Disable this rule in eslint config file

File: src/modules/auth/login/forms/login-form.tsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        <Button
+          type="submit"
+          variant="contained"
+          size="large"
+          color="info"
+          fullWidth={true}
+          disabled={status === 'pending'}
+          sx={styles.loginButton}
+        >
+          {status === 'pending' ? 'Logging in...' : 'Log in'}
```
Comment (NegatiV-64) [inline src/modules/auth/login/forms/login-form.tsx]: Create a special LoadingButton component or use from MUI

File: src/modules/auth/login/routes/login.route.tsx [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import Box from '@mui/material/Box';
+import { LoginForm } from '../forms/login-form';
+import { styles } from '../styles/styles';
```
Comment (NegatiV-64) [inline src/modules/auth/login/routes/login.route.tsx]: Please move the styles related to this component, to this component

File: src/modules/auth/login/types/auth-status.type.ts
Comment (NegatiV-64) [inline src/modules/auth/login/types/auth-status.type.ts:1]: Move this file to the shared level

File: src/modules/auth/login/types/auth.type.ts
Comment (NegatiV-64) [inline src/modules/auth/login/types/auth.type.ts:1]: Move this file to the shared level

File: src/modules/auth/login/types/current-user.type.ts
Comment (NegatiV-64) [inline src/modules/auth/login/types/current-user.type.ts:1]: Move it the auth type file

File: src/modules/auth/login/utils/get-local-user.util.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { type CurrentUser } from '../types/current-user.type';
+
+export function getLocalUser() {
+  const localUser = localStorage.getItem('tokens');
+
+  if (!localUser) {
+    return null;
+  }
+
+  return JSON.parse(localUser) as CurrentUser;
```
Comment (NegatiV-64) [inline src/modules/auth/login/utils/get-local-user.util.ts]: Validate the JSON parsed data

File: src/styles/root.css [hunk 13]
```diff
@@ -1,3 +1,5 @@
+@import url('https://fonts.googleapis.com/css2?family=Raleway:ital,wght@0,100..900;1,100..900&display=swap');
```
Comment (NegatiV-64) [inline src/styles/root.css]: MUI uses Roboto, let's keep use Roboto

Comment (NegatiV-64) [reply src/styles/root.css]: See: https://mui.com/material-ui/getting-started/installation/#roboto-font

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

File: src/modules/auth/types/login-form-data.type.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+export interface LoginFormData {
```
Comment (NegatiV-64) [inline src/modules/auth/types/login-form-data.type.ts]: Infer types from the schema of zod

File: src/shared/hooks/auth.hook.tsx [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    const localUser = getLocalUser();
+
+    if (localUser) {
+      dispatch(setCurrentUser(localUser));
+      dispatch(setStatus('success'));
+    } else {
+      dispatch(setStatus('idle'));
```
Comment (NegatiV-64) [inline src/shared/hooks/auth.hook.tsx]: if there's no user => anyway solve the status

File: src/shared/hooks/auth.hook.tsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  }, [dispatch]);
+
+  function login(credentials: LoginFormData) {
+    dispatch(setError(null));
+    dispatch(setStatus('pending'));
+
+    loginMutation.mutate(credentials, {
```
Comment (NegatiV-64) [inline src/shared/hooks/auth.hook.tsx]: Please move the API call to the login form submit:

- onSubmit in the form calls the login API => returns tokens
- onSubmit also handles errors during the API call
- Login function here accepts the tokens and just writes them into the localstorage and store. 

As a result, register page can reuse the same function

File: src/styles/root.css [hunk 17]
```diff
@@ -1,3 +1,5 @@
+@import url('https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap');
```
Comment (NegatiV-64) [inline src/styles/root.css]: [Use npm package](https://www.npmjs.com/package/@fontsource/roboto)

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

**Review by muhammadjon8 — DISMISSED** [review]

**Review by NegatiV-64 — APPROVED** [review]

**Pull request merged**